enhancements/machine-config/on-cluster-builds.md

```diff
@@ -0,0 +1,1009 @@
+  upgraded to a new OpenShift release, any custom OS content will be
+  upgraded as well (subject to update availability).
+
+- A cluster administrator will have the option to build and test their custom OS
+  image ahead of time and defer its deployment.
+
+- RHEL entitlements for custom OS content will be handled transparently
+  though there may be the requirement of a one-time setup step to
+  facilitate that.
+
+### Non-Goals
+
+- While we are enabling an OpenShift cluster admin to build a customized
+  OS image for their setup, we are not intending to implement a
+  general-purpose continuous integration (CI) system. That means there
+  will be intentional feature gaps around secrets management, testing
+  workflows, etc. If those features are required, it is recommended that
+  a cluster administrator use an off-cluster image build mechanism
+  instead.
+
+- There are many interesting conversations happening around integrating
+  image-based RHEL into OCP. A lot of these conversations have to do with
+  defining or redefining what promises RHCOS makes with regard to OS
+  immutability. While on-cluster builds must be aware of image-based RHEL
+  with that eventuality in mind, figuring out future OS immutability
+  capabilities is out-of-scope for this enhancement.
+
+- There will be situations where the MCO will need to work around feature gaps
+  in bootc. The MCO would not be alone since RHEL For Edge and RHIVOS are likely
+  to face similar situations and friction. While these concerns exist,
+  addressing them within the context of RHEL For Edge and RHIVOS is out-of-scope
+  for this enhancement.
+
+## Proposal
+
+### Workflow Description
+
+1.  The cluster administrator realizes that they need to add custom
+    content to the OS on each of their cluster nodes.
+
+2.  The cluster administrator edits the MachineConfigPool, adding a
+    Dockerfile, as well as the one-time setup of what image registry to push the
+    final image to, the credentials needed to push to this registry, etc.
+
+3.  The BuildController detects that these details are provided and
+    begins a build.
+
+4.  The built image is pushed to the cluster administrators’ image
+    registry of choice.
+
+5.  NodeController updates the `desiredImage` annotation on each node within its
+    pool according to its rules such as maximum unavailable nodes, which nodes
+    haven't been updated yet, etc.
+
+6.  The MachineConfigDaemon drains the target node, applies the new OS
+    content to the node, and reboots the node upon completion.
+
+#### Variation and form factor considerations
+
+##### Single-Node OpenShift (SNO)
+
+Single-Node OpenShift (SNO) should not have any difficulty running the
+on-cluster build process as-is. However, there may be some efficiency
+improvements that can be made. For example, the final image could be written
+directly to the nodes’ filesystem before the handoff to rpm-ostree / bootc to
```

Comment:
This could be deferred to a separate enhancement in the future.

Reply:
I'll add a note about that.